### DISCUSSION TITLE
Irritating Server to Client data passing

### DISCUSSION THREAD

**Discussion by igolkotek** [vikejs/vike#1376] [Discussion]
In fact here there're 2 different stories, related to `+onBeforeRender` and `+guard`, but they have the same Vike mechanism.

My presets are:
-  Client routing
-   onBeforeRender: {
           env: { server: true, client: true }
    }

According to the Vike idea, the server +onBeforeRender script passes data to the client (which it doesn't 100%), and does not run client side `+onBeforeRender` and `+guard` at the Client app launch.
Just the second page has the privilege to run them.

According to my concept I do not need a user authorization at the server side, but just at the browser one.   
But I can't do anything as both scripts are skipped.
I can do some workaround instead of onBeforeRender, but guard.js should be run both server and client side always, imho.

Thanks

**Comment by lourot** [comment]
It sounds like you need:
1. some things to be run on server or client interchangeably. Can you use a `data()` hook with `env: { server: true, client: true }` for this? It is brand new in v0.4.151 and is very similar to `onBeforeRender()`: https://vike.dev/data
2. some things to be run only on the server. Can you use an `onBeforeRender()` hook with `env: { server: true, client: false }` for this?

**Reply by igolkotek** [reply]
Thank you, I'll try.
But now my main headache is *guard* and the client only authentication.
As guard script is passed at the client startup, I have to find a workaround, and I don't like how it looks

**Reply by lourot** [reply]
Can you make guard `env: { server: false, client: true }`, so it always run on client?

**Reply by igolkotek** [reply]
yess, great!

**Reply by igolkotek** [reply]
Ooh,

making `meta`:
```
guard: {
    env: { server: false, client: true }
}
```
doesn't affect the client's 1st page hit - no log recorded. 
Just the same as  `{ server: true, client: true }`.   

The 2nd page requests `+guard.js` .

**Reply by lourot** [reply]
Do you have some [`pageContextInit`](https://vike.dev/pageContext) in your server code? We have [some logic](https://github.com/vikejs/vike/blob/main/vike/client/client-routing-runtime/getPageContextFromHooks.ts#L243) for running hooks on server despite the `env` because of `pageContextInit`.

If you have some `pageContextInit`, can you temporarily remove it in order to confirm that it's the reason why the hook gets run on server? If that's the case, we'll have to rethink that or maybe invent another kind of hook. If that's not the case, can you please push a reproducer so we can have a look? Thanks a lot.

**Reply by igolkotek** [reply]
Yes, I do, both in my project and test one.
After running simple `vite dev` , that is without server code, nothing has changed: the 1st page renderer doesn't run guard.

**Reply by lourot** [reply]
I can reproduce with https://github.com/vikejs/vike/tree/aurelien/reproducer/discussion-1376/examples/react-full

Let me have a look

**Reply by lourot** [reply]
I see. If we have either a `data()` or an `onBeforeRender()` hook that run on server, we refuse to run the `guard()` hook on the client, even if its env is client-only and even if we don't have a `pageContextInit`: https://github.com/vikejs/vike/blob/main/vike/client/client-routing-runtime/getPageContextFromHooks.ts#L176

In my reproducer there is a `data()` hook to be run on server. I believe we want to make sure the `guard()` hook runs before `data()`. So if `data()` runs on server first and we then later run `guard()` on the client, the sequence wouldn't be guaranteed.

@brillout do we really need to guarantee this sequence or could we relax it? Maybe we should document that at https://vike.dev/guard

@igolkotek as a workaround, could you use a `data()` or `onBeforeRender()` hook instead of a `guard()`? I believe you can achieve the same.

**Reply by brillout** [reply]
My plan was to make `guard()` server only by default, while enabling the user to change its env, just like tey can change the env of `data()` and `onBeforeEnv()`. I didn't read the whole thread but I think that would adress OP's issue (but let me know if you think I should read the whole threat).

**Reply by igolkotek** [reply]
@AurelienLourot @brillout 
Thank you for your explanations. My idea is to avoid using server guard in favour of client, as I'd prefer to have server cached response - the client could do authentication easily.
But the current vike process makes me to find some workaround.

**Reply by brillout** [reply]
> My plan was to make `guard()` server only by default, while enabling the user to change its env, just like tey can change the env of `data()` and `onBeforeEnv()`.

@igolkotek This is the plan but let us know if you see issues with that.

**Comment by igolkotek** [comment]
@AurelienLourot ,
> If we have either a `data()` or an `onBeforeRender()` hook that run on server, we refuse to run the `guard()` hook on the client, even if its env is client-only

It looks counter logical, doesn't it? 
As `server: false, client: true` should work if selected.

Sure, for myself I have a workaround (I hope :) ), but it looks to be good to change Vike code some day in the future.

**Reply by brillout** [reply]
@igolkotek Anything wrong with that [plan](https://github.com/vikejs/vike/discussions/1376#discussioncomment-8037119)? Wouldn't that resolve your situation in a neat way? Let me know if you believe you need anything else beyond that plan.

**Reply by brillout** [reply]
And, to be clear, once the plan is implemented, the sentence you just quoted wouldn't be true anymore.

**Comment by igolkotek** [comment]
> Anything wrong with that [plan](https://github.com/vikejs/vike/discussions/1376#discussioncomment-8037119)? Wouldn't that resolve your situation in a neat way? Let me know if you believe you need anything else beyond that plan.

@brillout 
I think there is something wrong with your plan realization.
Now vike has 3 special hooks:
- +data.js
- +onBeforeRender
- +guard

The first two are concerned with the data fetching, the last one - with authorization.
But because of them three just `+data.js` is really freely configured to be launched on server or browser, authorization issue of mine could be solved with data.js only which breaks down all the idea of concern separation.

And this leads us to misunderstanding of those hooks usage (and maybe you will realize the 4th one - `onStartup`).

Anyway I can almost solve my issue with +data.js, thank you )

**Reply by igolkotek** [reply]
The difference between data.js and guard.js for authorization is that data.js must be in the every url of concern, but guard.js should be located at the root of a nested url tree just once.

So data.js is a workaround, but it looks right to allow guard be launched on server or browser according to an app business logic.

**Reply by brillout** [reply]
I'm not sure I undersand what you're saying. But I'm glad you've found a workaround; feel free to let us know if you run into a blocker.

**Reply by igolkotek** [reply]
OK, thank you for your help, I'm finishing this discussion.

Just a few notes.

As I see you have a small team and are in a permanent hurry with the project. I'm so impressed you have time to discuss issues and help. Many thanks.. 

Vike which looks like a "simple" server middleware, in fact is a complicated project of 2 different situations: serverRouting and clientRouting.

Hope I'm wrong but imho your project lacks the clear vision of it's growth.

You offer 2 ways to separate code logic:
- you added `.client.js` and `browser.js` , but they have no affect on `+guard.client(server).js` or `+onBeforeRender.client(server).js` .
- you added env: {server, client}, but they have no effect for `guard` and `onBeforeRender`. 
 
Maybe I just do not understand it, but for me it looks embarrassing.

**Reply by brillout** [reply]
I agree, and using `.client.js` and `.server.js` for changing `env` is actually very much a consideration. But it isn't the highest priority right now.

As for Server Routing and Client Routing, there are valid use cases for both. That said, we do plan to make Client Routing the default in the documentation, in order to make the docs easier for most users which will be using Client Routing.

**Reply by igolkotek** [reply]
Thank you )
